File: sample-videochat-webrtc/sample-videochat-webrtc/Components/Call/CallHelper/CallPayload.m
//
//  CallPayload.m
//  sample-videochat-webrtc
//
//  Created by Injoit on 11.09.2020.
//  Copyright © 2020 QuickBlox Team. All rights reserved.
//

#import "CallPayload.h"
#import "Profile.h"
#import "NSDate+Videochat.h"

#define callInfoKey( prop ) NSStringFromSelector(@selector(prop))

@interface CallPayload ()

@property (nonatomic, strong) NSString *opponentsIDs;
@property (nonatomic, strong) NSString *sessionID;
@property (nonatomic, strong) NSString *contactIdentifier;
@property (nonatomic, strong) NSString *timestamp;
@property (nonatomic, strong) NSString *conferenceType;

@property (nonatomic, strong) Profile *currentUser;

@end

@implementation CallPayload

- (NSString *)title {
    if (self.members.allValues.count) {
        return [self.members.allValues componentsJoinedByString: @", "];
    }
    return self.contactIdentifier;
}

- (BOOL)valid {
    return self.sessionID.length;
}

- (BOOL)missed {
    if (self.members.count < 1) {
        return YES;
    }
    
    if (!self.timestamp.length) {
        return YES;
    }

    NSTimeInterval startCallTimeInterval = self.timestamp.longLongValue;
    if (startCallTimeInterval) {
        NSTimeInterval timeNow = [[NSDate date] currentTimestamp];
        return (timeNow - startCallTimeInterval) / 1000 > QBRTCConfig.answerTimeInterval;
    }
    
    return NO;
}

- (BOOL)hasVideo {
    return [self.conferenceType isEqualToString:@"1"];
}

- (instancetype)initWithPayload:(NSDictionary *)payload {
    self = [super init];
    if (self) {
        self.opponentsIDs = payload[callInfoKey(opponentsIDs)] ?: @"";
        self.sessionID = payload[callInfoKey(sessionID)] ?: @"";

        NSString *contactIdentifier = payload[callInfoKey(contactIdentifier)];
        NSArray<NSString *>*participantsNames;
        if (contactIdentifier.length) {
            self.contactIdentifier = contactIdentifier;
            participantsNames = [self.contactIdentifier componentsSeparatedByString:@","];
        } else {
            self.contactIdentifier = @"Incoming call. Connecting...";
            participantsNames = [self.opponentsIDs componentsSeparatedByString:@","];
        }
        
        self.timestamp = payload[callInfoKey(timestamp)] ?: @"";
        self.conferenceType = payload[callInfoKey(conferenceType)] ?: @"1";
        
        self.currentUser = [[Profile alloc] init];
        
        _members = @{};
        NSArray<NSString *>*participantsIDs = [self.opponentsIDs componentsSeparatedByString:@","];
        
        if (participantsIDs.count == participantsNames.count) {
            NSMutableArray *ids = @[].mutableCopy;
            NSNumberFormatter *formatter = [NSNumberFormatter new];
            formatter.numberStyle = NSNumberFormatterNoStyle;
            for (NSString *stringId in participantsIDs) {
                [ids addObject:[formatter numberFromString:stringId]];
            }
            NSMutableDictionary *participants =
            [NSMutableDictionary dictionaryWithObjects:participantsNames
                                               forKeys:ids.copy];
            [participants removeObjectForKey:@(self.currentUser.ID)];
            if (participants.count) { _members = participants; }
        }
    }
    return self;
}

@end

File: sample-videochat-webrtc-swift/sample-videochat-webrtc-swift/Components/Calls/CallHelper/CallPayload.swift
//
//  CallPayload.swift
//  sample-videochat-webrtc-swift
//
//  Created by Injoit on 01.04.2021.
//  Copyright © 2021 QuickBlox. All rights reserved.
//

import UIKit
import QuickbloxWebRTC

class CallPayload: NSObject {
    //MARK: - Properties
    /// Some call data wrong or absent
    var valid: Bool {
        return sessionID.isEmpty == false
    }
    
    var missed = false
    
    var hasVideo: Bool {
        return conferenceType == "1"
    }
    
    private(set) var sessionID: String = ""
    /// The call participants without a current user. The key is a user id and the value is a user name.
    private(set) var members: [NSNumber: String] = [:]
    var title: String {
        if members.values.isEmpty == false {
            return members.values.joined(separator: ", ")
        }
        return contactIdentifier
    }
    private var currentUser = Profile()
    
    private var opponentsIDs: String = ""
    private var contactIdentifier: String = ""
    var timestamp: String = ""
    private var conferenceType: String = "1"
    
    //MARK: - Life Cycle
    init(payload: [String : String]) {
        super.init()
        
        opponentsIDs = payload["opponentsIDs"] ?? ""
        var participantsNames: [String] = []
        if let contactIdentifier = payload["contactIdentifier"] {
            self.contactIdentifier = contactIdentifier
            participantsNames = contactIdentifier.components(separatedBy: ",")
        } else {
            contactIdentifier = "Incoming call. Connecting..."
            participantsNames = opponentsIDs.components(separatedBy: ",")
        }
        sessionID = payload["sessionID"] ?? ""
        conferenceType = payload["conferenceType"] ?? "1"
        timestamp = payload["timestamp"] ?? ""
        currentUser = Profile()
        guard currentUser.isFull == true else {
            return
        }
        let participantsIDs = opponentsIDs.components(separatedBy: ",")
        if participantsIDs.count == participantsNames.count {
            let ids = participantsIDs.compactMap({NSNumber(value: Int($0)!)})
            var participants: [NSNumber: String] = [:]
            for i in 0...ids.count - 1 {
                participants[ids[i]] = participantsNames[i]
            }
            participants.removeValue(forKey: NSNumber(value: currentUser.ID))
            if participants.isEmpty == false {
                members = participants
            }
        }
        
        if members.count < 1 {
            missed = true
            return
        }
        let startTimeInterval = Int64(timestamp) ?? Date().timeStamp
        let timeIntervalNow = Date().timeStamp
        let receivedTimeInterval = (timeIntervalNow - startTimeInterval) / 1000
        missed = receivedTimeInterval > Int64(QBRTCConfig.answerTimeInterval())
    }
}
